README


This archive contains the distribution Role-REST-Client,
version 0.18:

  REST Client Role

This software is copyright (c) 2014 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.015.



Changes
Revision history for Role-REST-Client

0.18    2014-06-26 22:28:12 Europe/Copenhagen
 - Decode the content even for http codes >= 400. There might be some information there (idea: moltar)

0.17    2014-04-22 19:58:08 Europe/Copenhagen
 - Prevent 'redefined' warnings and updated dependency list (Breno G. de Oliveira)

0.16    2013-09-26 07:51:17 Europe/Copenhagen
 - Mooify Role::REST::Client (Matt Phillips)
 - Start deserializing text/javascript without a warning (Mark Stosberg)
 - Replace HTTP::Tiny with HTTP::Thin
 - Use HTTP::Response to determine if a request failed

0.15    2013-03-28 22:38:29 Europe/Copenhagen
 - Only update httpheaders from persistent_headers if they are not already set.

0.14    2013-03-28 21:17:06 Europe/Copenhagen

 - Fix URI encode process (Wallace Reis)
 - self->_persistent_headers was sometimes undef and there was no lazy bit.
 - dist.ini corrections
 - Improved Docs (Mark Stosberg) 

0.13    2012-11-11 12:45:40 Europe/Copenhagen
 -  Change json/yaml/xml serializers to recommends instead of requires (Matt Phillips)
 -  Fix httpheaders attr - must be lazy since builder depends on persistent_headers (Wallace Reis).
 -  Remove MooseX::Declare from the dependency chain (Mark Stosberg)
 -  Don't require HTTP::Tiny to be loaded if it is never used (Mark Stosberg)

0.12    2012-06-26 19:08:09 Europe/Copenhagen
 -  Fix httpheaders attr - must be lazy since builder depends on persistent_headers (Wallace Reis).

0.11    2012-06-14 20:47:25 Europe/Copenhagen
 - Init methods remember the persistent headers (Thanks to Wallace Reis for the tests).

0.10    2012-06-01 14:07:37 Europe/Copenhagen
 - Add HTTP HEAD method (Wallace Reis).

0.09    2012-05-29 14:59:38 Europe/Copenhagen
 - Use Data::Serializer::Raw instead of Data::Serializer
 - Fix content-length missmatch issue (Wallace Reis).
 - Factor out request step (Wallace Reis).

0.08    2012-05-29 07:24:24 Europe/Copenhagen
 - Lazily deserialize the response (Wallace Reis).
 - Allow use of a custom UA object (Wallace Reis).
 - Defer the building of serializer obj until really necessary (Wallace Reis).
 
0.07    2012-05-24 07:22:16 Europe/Copenhagen
 - Make new_serializer and serializer_class public method and attr,
   respectively (Wallace Reis).

0.06    2012-05-19 18:32:07 Europe/Copenhagen
 - Add preserve_headers option and persistent_headers to allow sticky headers.

0.05    2012-04-02 22:57:55 Europe/Copenhagen
 - Response and serializer are now easy overridable [Matt Phillips].

0.04    2012-03-30 11:02:47 Europe/Copenhagen
 - Don't use 5.010; we don't have particular needs for that

0.03    2012-03-19 19:50:25 Europe/Copenhagen
 - Clear headers AFTER the request, Makes it possible to set any http header except content-type
   and any header set by authentication roles.

0.02    2012-03-01 20:58:30 Europe/Copenhagen
 - Turn _ua into an attribute
 - Add support for Auth modules (See http://search.cpan.org/dist/Role-REST-Client-Auth-Basic/)

0.01    2012-02-20 17:28:59 Europe/Copenhagen
 - First version, with functionality taken from Catalyst::Model::REST


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.015.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "REST Client Role",
  "AUTHOR" => "Kaare Rasmussen <kaare at cpan dot net>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Role-REST-Client",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Role::REST::Client",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Serializer::Raw" => 0,
    "HTTP::Headers" => 0,
    "HTTP::Message" => 0,
    "HTTP::Response" => 0,
    "HTTP::Status" => 0,
    "HTTP::Thin" => 0,
    "HTTP::Tiny" => 0,
    "Moo" => "1.003000",
    "Moo::Role" => 0,
    "MooX::HandlesVia" => "0.001004",
    "Try::Tiny" => 0,
    "Type::Tiny" => "0.024",
    "Types::Standard" => 0,
    "URI::Escape::XS" => "0.11"
  },
  "TEST_REQUIRES" => {
    "Data::Serializer::JSON" => 0,
    "Data::Serializer::XML::Simple" => 0,
    "Data::Serializer::YAML" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "HTTP::Request" => 0,
    "HTTP::Thin" => 0,
    "JSON" => "2.00",
    "Moose" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "XML::Simple" => 0,
    "YAML" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.18",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Data::Serializer::JSON" => 0,
  "Data::Serializer::Raw" => 0,
  "Data::Serializer::XML::Simple" => 0,
  "Data::Serializer::YAML" => 0,
  "File::Find" => 0,
  "File::Temp" => 0,
  "HTTP::Headers" => 0,
  "HTTP::Message" => 0,
  "HTTP::Request" => 0,
  "HTTP::Response" => 0,
  "HTTP::Status" => 0,
  "HTTP::Thin" => 0,
  "HTTP::Tiny" => 0,
  "JSON" => "2.00",
  "Module::Build" => "0.3601",
  "Moo" => "1.003000",
  "Moo::Role" => 0,
  "MooX::HandlesVia" => "0.001004",
  "Moose" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "Try::Tiny" => 0,
  "Type::Tiny" => "0.024",
  "Types::Standard" => 0,
  "URI::Escape::XS" => "0.11",
  "XML::Simple" => 0,
  "YAML" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Role/REST/Client.pm
package Role::REST::Client;
$Role::REST::Client::VERSION = '0.18';
use Moo::Role;
use MooX::HandlesVia;
use Types::Standard qw(HashRef Str Int Enum HasMethods);

use HTTP::Tiny;
use URI::Escape::XS 'uri_escape';
use Try::Tiny;
use Carp qw(confess);
use HTTP::Response;
use HTTP::Status 'status_message';
use HTTP::Headers;

use Role::REST::Client::Serializer;
use Role::REST::Client::Response;

has 'server' => (
	isa => Str,
	is  => 'rw',
);

has 'type' => (
	isa => Enum[qw{application/json application/xml application/yaml application/x-www-form-urlencoded}],
	is  => 'rw',
	default => sub { 'application/json' },
);

has clientattrs => (
	isa => HashRef,
	is => 'ro',
	default => sub {return {} }
);

has user_agent => (
	isa => HasMethods['request'],
	is => 'ro',
	lazy => 1,
	builder => '_build_user_agent',
);

sub _build_user_agent {
	my $self = shift;
	require HTTP::Thin;
	return HTTP::Thin->new(%{$self->clientattrs});
}

has persistent_headers => (
	is        => 'lazy',
#	isa       => HashRef[Str],
	default   => sub { {} },
	trigger   => sub {
		my ( $self, $header, $old_header ) = @_;
		# Update httpheaders if their value was initialized first
		while (my ($key, $value) = each %$header) {
			$self->set_header($key, $value) unless $self->exist_header($key);
		}
	},
	handles_via => 'Hash',
	handles   => {
		set_persistent_header     => 'set',
		get_persistent_header     => 'get',
		has_no_persistent_headers => 'is_empty',
		clear_persistent_headers  => 'clear',
	},
);

has httpheaders => (
	is          => 'lazy',
	isa         => HashRef[Str],
	writer      => '_set_httpheaders',
	handles_via => 'Hash',
	handles     => {
		set_header     => 'set',
		get_header     => 'get',
		exist_header   => 'exists',
		has_no_headers => 'is_empty',
		clear_headers  => 'clear',
	},
);

has serializer_class => (
	isa => Str,
	is => 'ro',
	default => sub { 'Role::REST::Client::Serializer' },
);

sub _build_httpheaders {
	my ($self, $headers) = @_;
	$headers ||= {};
	$self->_set_httpheaders( { %{$self->persistent_headers}, %$headers });
}

sub reset_headers {my $self = shift;$self->_set_httpheaders({ %{$self->persistent_headers} })}

sub _rest_response_class { 'Role::REST::Client::Response' }

# If the response is a hashref, we expect it to be in the format returned by
# HTTP::Tiny->request() and convert it to an HTTP::Response object.  Otherwise,
# pass the response through unmodified.
sub _handle_response {
	my ( $self, $res ) = @_;
	if ( ref $res eq 'HASH' ) {
		my $code = $res->{'status'};
		return HTTP::Response->new(
			$code,
			$res->{'reason'} || status_message($code),
			HTTP::Headers->new(%{$res->{'headers'}}),
			$res->{'content'},
		);
	} else {
		return $res;
	}
}

sub _new_rest_response {
	my ($self, @args) = @_;
	return $self->_rest_response_class->new(@args);
}

sub new_serializer {
	my ($self, @args) = @_;
	$self->serializer_class->new(@args);
}

sub _serializer {
	my ($self, $type) = @_;
	$type ||= $self->type;
	$type =~ s/;\s*?charset=.+$//i; #remove stuff like ;charset=utf8
	try {
		$self->{serializer}{$type} ||= $self->new_serializer(type => $type);
	}
	catch {
		# Deal with real life content types like "text/xml;charset=ISO-8859-1"
		warn "No serializer available for " . $type . " content. Trying default " . $self->type;
		$self->{serializer}{$type} = $self->new_serializer(type => $self->type);
	};
	return $self->{serializer}{$type};
}

sub do_request {
	my ($self, $method, $uri, $opts) = @_;
	return $self->user_agent->request($method, $uri, $opts);
}

sub _call {
	my ($self, $method, $endpoint, $data, $args) = @_;
	my $uri = $self->server.$endpoint;
	# If no data, just call endpoint (or uri if GET w/parameters)
	# If data is a scalar, call endpoint with data as content (POST w/parameters)
	# Otherwise, encode data
	$self->set_header('content-type', $self->type);
	my %options = (headers => $self->httpheaders);
	if ( defined $data ) {
		$options{content} = ref $data ? $self->_serializer->serialize($data) : $data;
		$options{'headers'}{'content-length'} = length($options{'content'});
	}
	my $res = $self->_handle_response( $self->do_request($method, $uri, \%options) );

	$self->reset_headers unless $args->{preserve_headers};
	my $deserializer_cb = sub {
		# Try to find a serializer for the result content
		my $content_type = $args->{deserializer} || $res->header('Content-Type');
		my $deserializer = $self->_serializer($content_type);
		# Try to deserialize
		my $content = $res->decoded_content;
		$content = $deserializer->deserialize($content) if $deserializer && $content;
		$content ||= {};
	};
	return $self->_new_rest_response(
		code => $res->code,
		response => $res,
		data => $deserializer_cb,
		$res->is_error ? ( error => $res->message) : (),
	);
}

sub _urlencode_data {
	my ($self, $data) = @_;
	return join '&', map { uri_escape($_) . '=' . uri_escape($data->{$_})} keys %$data;
}

sub _request_with_query {
	my ($self, $method, $endpoint, $data, $args) = @_;
	my $uri = $endpoint;
	if ($data && scalar keys %$data) {
		$uri .= '?' . $self->_urlencode_data($data);
	}
	return $self->_call($method, $uri, undef, $args);
}

sub get { return shift->_request_with_query('GET', @_) }

sub head { return shift->_request_with_query('HEAD', @_) }

sub delete { return shift->_request_with_query('DELETE', @_) }

sub _request_with_body {
	my ($self, $method, $endpoint, $data, $args) = @_;
	my $content = $data;
	if ( $self->type =~ /urlencoded/ ) {
		$content = ( $data && scalar keys %$data ) ? $self->_urlencode_data($data) : q{};
	}
	return $self->_call($method, $endpoint, $content, $args);
}

sub post { return shift->_request_with_body('POST', @_) }

sub put { return shift->_request_with_body('PUT', @_) }

sub options { return shift->_request_with_body('OPTIONS', @_) }

1;

=pod

=encoding UTF-8

=head1 NAME

Role::REST::Client - REST Client Role

=head1 VERSION

version 0.18

=head1 SYNOPSIS

	{
		package RESTExample;

		use Moose;
		with 'Role::REST::Client';

		sub bar {
			my ($self) = @_;
			my $res = $self->post('/foo/bar/baz', {foo => 'bar'});
			my $code = $res->code;
			my $data = $res->data;
			return $data if $code == 200;
	   }

	}

	my $foo = RESTExample->new(
		server =>      'http://localhost:3000',
		type   =>      'application/json',
		clientattrs => {timeout => 5},
	);

	$foo->bar;

	# controller
	sub foo : Local {
		my ($self, $c) = @_;
		my $res = $c->model('MyData')->post('/foo/bar/baz', {foo => 'bar'});
		my $code = $res->code;
		my $data = $res->data;
		...
	}

=head1 DESCRIPTION

This REST Client role makes REST connectivety easy.

Role::REST::Client will handle encoding and decoding when using the HTTP verbs.

	GET
	PUT
	POST
	DELETE
	OPTIONS
	HEAD

Currently Role::REST::Client supports these encodings

	application/json
	application/x-www-form-urlencoded
	application/xml
	application/yaml

x-www-form-urlencoded only works for GET and POST, and only for encoding, not decoding.

=head1 NAME

Role::REST::Client - REST Client Role

=head1 METHODS

=head2 methods

Role::REST::Client implements the standard HTTP 1.1 verbs as methods

	post
	get
	head
	put
	delete
	options

All methods take these parameters

	url - The REST service
	data - The data structure (hashref, arrayref) to send. The data will be encoded
		according to the value of the I<type> attribute.
	args - hashref with arguments to augment the way the call is handled.

args - the optional argument parameter can have these entries

	deserializer - if you KNOW that the content-type of the response is incorrect,
	you can supply the correct content type, like
	my $res = $self->post('/foo/bar/baz', {foo => 'bar'}, {deserializer => 'application/yaml'});

	preserve_headers - set this to true if you want to keep the headers between calls

All methods return a response object dictated by _rest_response_class. Set to L<Role::REST::Client::Response> by default.

=head1 ATTRIBUTES

=head2 user_agent

  sub _build_user_agent { HTTP::Thin->new }

A User Agent object which has a C<< ->request >> method suitably compatible with L<HTTP::Tiny>. It should accept arguments like this: C<< $ua->request($method, $uri, $opts) >>, and needs to return a hashref as HTTP::Tiny does, or an L<HTTP::Response> object.  To set your own default, use a C<_build_user_agent> method.

=head2 server

URL of the REST server.

e.g. 'http://localhost:3000'

=head2 type

MIME Content-Type header,

e.g. application/json

=head2 httpheaders

  $self->set_header('Header' => 'foo', ... );
  $self->get_header('Header-Name');
  $self->has_no_headers;
  $self->clear_headers;

You can set any http header you like with set_header, e.g.
$self->set_header($key, $value) but the content-type header will be overridden.

=head2 persistent_headers

  $self->set_persistent_header('Header' => 'foo', ... );
  $self->get_persistent_header('Header-Name');
  $self->has_no_persistent_headers;
  $self->clear_persistent_headers;

A hashref containing headers you want to use for all requests. Use the methods
described above to manipulate it.

To set your own defaults, override the default or call C<set_persistent_header()> in your
C<BUILD> method.

  has '+persistent_headers' => (
    default => sub { ... },
  );

=head2 clientattrs

Attributes to feed the user agent object (which defaults to L<HTTP::Thin>)

e.g. {timeout => 10}

=head1 CONTRIBUTORS

Breno G. de Oliveira, <garu@cpan.org>

Mark Stosberg, <mark@stosberg.com>

Matt Phillips, (cpan:MATTP) <mattp@cpan.org>

Wallace Reis, <wallace@reis.me>

=head1 BUGS

Please report any bugs or feature requests to bug-role-rest-client at rt.cpan.org, or through the
web interface at http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Role-REST-Client.

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

__END__

# ABSTRACT: REST Client Role



lib/Role/REST/Client/Response.pm
package Role::REST::Client::Response;
$Role::REST::Client::Response::VERSION = '0.18';
use Moo;
use MooX::HandlesVia;
use Types::Standard qw(Str Int CodeRef InstanceOf);

has 'code' => (
	isa => Int,
	is  => 'ro',
);
has 'response' => (
	isa => InstanceOf['HTTP::Response'],
	is  => 'ro',
);
has 'error' => (
	isa => Str,
	is  => 'ro',
	predicate => 'failed',
);
has 'data_callback' => (
	init_arg => 'data',
	isa => CodeRef,
	is  => 'ro',
	default => sub { sub { {} } },
	handles_via => 'Code',
	handles => { data => 'execute' },
);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Role::REST::Client::Response

=head1 VERSION

version 0.18

=head1 SYNOPSIS

    my $res = Role::REST::Client::Response->new(
        code          => '200',
        response      => HTTP::Response->new(...),
        error         => 0,
        data_callback => sub { sub { ... } },
    );

=head1 NAME

Role::REST::Client::Response - Response class for REST

=head1 ATTRIBUTES

=head2 code

HTTP status code of the request

=head2 response

L<HTTP::Response> object. Use this if you need more information than status and content.

=head2 error

The error description returned from the user agent when the HTTP status code is 500 or higher. More detail may be found 
by calling C<< $res->response->content >>.

=head2 failed

True if the request didn't succeed.

=head2 data

The deserialized data. Returns an empty hashref if the response was unsuccessful.

=head1 BUGS

Please report any bugs or feature requests to bug-role-rest-client at rt.cpan.org, or through the
web interface at http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Role-REST-Client.

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Role/REST/Client/Serializer.pm
package Role::REST::Client::Serializer;
$Role::REST::Client::Serializer::VERSION = '0.18';
use Try::Tiny;
use Moo;
use Types::Standard qw(Enum InstanceOf);
use Data::Serializer::Raw;

has 'type' => (
	isa => Enum[qw{application/json application/xml application/yaml application/x-www-form-urlencoded text/javascript}],
	is  => 'rw',
	default => sub { 'application/json' },
);

has 'serializer' => (
	isa => InstanceOf['Data::Serializer::Raw'],
	is => 'ro',
	default => \&_set_serializer,
	lazy => 1,
);

our %modules = (
	'application/json' => {
		module => 'JSON',
	},
	'application/xml' => {
		module => 'XML::Simple',
	},
	'application/yaml' => {
		module => 'YAML',
	},
	'application/x-www-form-urlencoded' => {
		module => 'FORM',
	},
	'text/javascript' => {
	        module => 'JSON',	
	},
);

sub _set_serializer {
	my $self = shift;
	return unless $modules{$self->type};

	my $module = $modules{$self->type}{module};
	return $module if $module eq 'FORM';

	return Data::Serializer::Raw->new(
		serializer => $module,
	);
}

sub content_type {
	my ($self) = @_;
	return $self->type;
}

sub serialize {
	my ($self, $data) = @_;
	return unless $self->serializer;

	my $result;
	try {
		$result = $self->serializer->serialize($data)
	} catch {
		warn "Couldn't serialize data with " . $self->type;
	};

	return $result;
}

sub deserialize {
	my ($self, $data) = @_;
	return unless $self->serializer;

	my $result;
	try {
		$result = $self->serializer->deserialize($data);
	} catch {
		warn "Couldn't deserialize data with " . $self->type;
	};

	return $result;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Role::REST::Client::Serializer

=head1 VERSION

version 0.18

=head1 AUTHOR

Kaare Rasmussen <kaare at cpan dot net>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Kaare Rasmussen.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Build.PL

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.015.
use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "REST Client Role",
  "dist_author" => [
    "Kaare Rasmussen <kaare at cpan dot net>"
  ],
  "dist_name" => "Role-REST-Client",
  "dist_version" => "0.18",
  "license" => "perl",
  "module_name" => "Role::REST::Client",
  "recommends" => {
    "Data::Serializer::JSON" => 0,
    "Data::Serializer::XML::Simple" => 0,
    "Data::Serializer::YAML" => 0,
    "HTTP::Thin" => 0,
    "JSON" => "2.00",
    "XML::Simple" => 0,
    "YAML" => 0
  },
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Serializer::Raw" => 0,
    "HTTP::Headers" => 0,
    "HTTP::Message" => 0,
    "HTTP::Response" => 0,
    "HTTP::Status" => 0,
    "HTTP::Thin" => 0,
    "HTTP::Tiny" => 0,
    "Moo" => "1.003000",
    "Moo::Role" => 0,
    "MooX::HandlesVia" => "0.001004",
    "Try::Tiny" => 0,
    "Type::Tiny" => "0.024",
    "Types::Standard" => 0,
    "URI::Escape::XS" => "0.11"
  },
  "script_files" => [],
  "test_requires" => {
    "Data::Serializer::JSON" => 0,
    "Data::Serializer::XML::Simple" => 0,
    "Data::Serializer::YAML" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "HTTP::Request" => 0,
    "HTTP::Thin" => 0,
    "JSON" => "2.00",
    "Moose" => 0,
    "Test::Deep" => 0,
    "Test::More" => "0.88",
    "XML::Simple" => 0,
    "YAML" => 0,
    "strict" => 0,
    "utf8" => 0,
    "warnings" => 0
  }
);


my %fallback_build_requires = (
  "Data::Serializer::JSON" => 0,
  "Data::Serializer::XML::Simple" => 0,
  "Data::Serializer::YAML" => 0,
  "File::Find" => 0,
  "File::Temp" => 0,
  "HTTP::Request" => 0,
  "HTTP::Thin" => 0,
  "JSON" => "2.00",
  "Module::Build" => "0.3601",
  "Moose" => 0,
  "Test::Deep" => 0,
  "Test::More" => "0.88",
  "XML::Simple" => 0,
  "YAML" => 0,
  "strict" => 0,
  "utf8" => 0,
  "warnings" => 0
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
